### PR TITLE
Rectify async checks logic in presence of AsyncMethodBuilder attribute

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/60332

### CODE REVIEW

**Review by Youssef1313 — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs [hunk 1]
```diff
@@ -1102,7 +1102,7 @@ protected void AsyncMethodChecks(bool verifyReturnType, Location errorLocation,
 
                 if (this.HasAsyncMethodBuilderAttribute(out _))
                 {
-                    hasErrors |= MessageID.IDS_AsyncMethodBuilderOverride.CheckFeatureAvailability(diagnostics, this.DeclaringCompilation, errorLocation);
+                    hasErrors |= !MessageID.IDS_AsyncMethodBuilderOverride.CheckFeatureAvailability(diagnostics, this.DeclaringCompilation, errorLocation);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:1105]: >!MessageID.IDS_AsyncMethodBuilderOverride.CheckFeatureAvailability(diagnostics, this.DeclaringCompilation, errorLocation); [](http://example.com/codeflow?start=33&length=122)

In my opinion registering language version check in ```hasErrors``` is not the right thing to do. Wrong language version shouldn't change the analysis, and here, it clearly does.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:1105]: In that case, should https://github.com/dotnet/roslyn/blob/9c04fee59aae2c7faa8b02cf08bb3bcc9d637664/src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs#L226-L230 be tracked in a new issue to update as well?

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTasklikeMoreTests.cs [hunk 2]
```diff
@@ -1594,5 +1594,72 @@ public static void M(MyResult r)
...
+                [AsyncMethodBuilder(typeof(PoolingAsyncValueTaskMethodBuilder))]
+                static async ValueTask m1(ref int x, out int y) // 2
+                {
+                    await Task.Delay(1000);
+                    y = x * x;
+                    x = y * y;
+                }
+                """;
+
+            var compilation = CreateCompilation(source, parseOptions: useCSharp9 ? TestOptions.Regular9 : TestOptions.Regular10, targetFramework: TargetFramework.Net70);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTasklikeMoreTests.cs:1625]: >Regular10 [](http://example.com/codeflow?start=118&length=9)

Regular?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs]: @Youssef1313 That seems a good idea. There's at least one check there that is conditional on `hasErrors`.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs]: Filed https://github.com/dotnet/roslyn/issues/65363

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTasklikeMoreTests.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTasklikeMoreTests.cs:1625]: I think Regular10 is better. That's what we usually do for explicit LangVer tests that verify the boundary when the feature was introduced.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review (one-line fix). Thanks

**Review by cston — APPROVED** [review]

**Pull request merged**